### PR TITLE
Teacher Tool: Variable Usage Validator

### PR DESCRIPTION
**Pull request by thsparks**
This adds a new validator which can check if variables are both defined and used. The key difference between this and "block-exist" checks for `variable_set` and `variable_get` is that we can ensure the names match, and we can ensure it's checking for multiple variables (so `variable_get` isn't just called X number of times on the same variable, even if `variable_set` defines X unique variables, and so on...).

You can optionally filter by name, but we don't expose this currently in any criteria. You can also used nested validation to ensure a variable is set to something specific (like `pick random`), but again, we don't have any criteria that currently use this.

### CODE REVIEW

File: common-docs/teachertool/catalog-shared.json [hunk 1]
```diff
@@ -83,6 +67,22 @@
...
                 }
             ]
+        },
+        {
+            "id": "0DFA44C8-3CA5-4C77-946E-AF09F6C03879",
+            "use": "variable_usage",
+            "template": "Uses at least ${count} variables",
+            "docPath": "/teachertool",
+            "description": "The program creates and uses at least this many user-defined variables.",
+            "maxCount": 1,
```
[the remark this answers is not in the document]
Comment (thsparks) [reply common-docs/teachertool/catalog-shared.json:77]: We would, but this criteria doesn't have a name parameter. That's functionality in the validator but we don't expose it yet.

File: pxteditor/code-validation/validateVariableUsage.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+} {
+    const varDefinitionBlocks: Map<string, Blockly.Block[]> = new Map();
+    const usedVars: Set<string> = new Set();
+
+    for (const block of usedBlocks) {
+        if (!block.isEnabled()) {
+            continue;
+        }
+
+        const varsUsed = block.getVarModels();
```
[the remark this answers is not in the document]
Comment (thsparks) [reply pxteditor/code-validation/validateVariableUsage.ts:27]: getVars just returns string of variable ids, but getVarModels returns full objects with type information (which we don't set) alongside the name and a function to get the id: https://developers.google.com/blockly/reference/js/blockly.variablemodel_class

File: common-docs/teachertool/catalog-shared.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (srietkerk) [reply common-docs/teachertool/catalog-shared.json:77]: Ah, I see, makes sense.

File: pxteditor/code-validation/validateVariableUsage.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (srietkerk) [reply pxteditor/code-validation/validateVariableUsage.ts:27]: Awesome, thank you!

**Review by srietkerk — APPROVED** [review]
Looks good!

**Pull request merged**